docs/fsharp/language-reference/byrefs.md

```diff
@@ -0,0 +1,243 @@
+
+A "`byref`-like" struct in F# is a stack-bound value type. It is never allocated on the managed heap. They are useful for high-performance programming, as they are enforced with set of strong checks about the lifetimes and non-capture. These rules are:
+
+* They can be used as function parameters, method parameters, local variables, method returns.
+* They cannot be static or instance members of a class or normal struct.
+* They cannot be captured by any closure construct (`async` methods or lambda expressions).
+* They cannot be used as a generic parameter.
+
+This last point is crucial for F# "pipeline" style programming, as `|>` is a generic function that parameterizes its input types. This restriction may be relaxed for `|>` in the future, as it is inline and does not make any calls to non-inlined generic functions in its body.
+
+Although these rules very strongly restrict usage, they do so to fulfill the promise of high-performance computing in a safe manner.
+
+## Byref returns
+
+Byref returns from F# functions or members can be produced and consumed. When consuming a `byref`-returning method, the value is implicitly dereferenced. For example:
+
+```fsharp
+let safeSum(bytes: Span<byte>) =
+    let mutable sum = 0
+    for i in 0 .. bytes.Length - 1 do
+        sum <- sum + int bytes.[i]
+    sum
+
+let sum = safeSum(mySpanOfBytes)
+printfn "%d" sum // 'sum' is of type 'int'
+```
+
+To avoid the implicit dereference, such as passing a reference through multiple chained calls, use `&x` (where `x` is the value).
+
+You can also directly assign to a return `byref`. Consider the following (highly imperative) program:
+
+```fsharp
+ype C() =
+    let mutable nums = [| 1; 3; 7; 15; 31; 63; 127; 255; 511; 1023 |]
+
+    override __.ToString() = String.Join(' ', nums)
+
+    member __.FindLargestSmallerThan(target: int) =
+        let mutable ctr = nums.Length - 1
+
+        while ctr > 0 && nums.[ctr] >= target do ctr <- ctr - 1
+
+        if ctr > 0 then &nums.[ctr] else &nums.[0]
+
+[<EntryPoint>]
+let main argv =
+    let c = C()
+    printfn "Original sequence: %s" (c.ToString())
+
+    let v = &c.FindLargestSmallerThan 16
+
+    v <- v*2 // Directly assign to the byref return
+
+    printfn "New sequence:      %s" (c.ToString())
+
+    0 // return an integer exit code
+```
+
+This is the output:
+
+```console
+Original sequence: 1 3 7 15 31 63 127 255 511 1023
+New sequence:      1 3 7 30 31 63 127 255 511 1023
+```
+
+## Scoping for byrefs
+
+A `let`-bound value cannot have its reference escape the scope in which it was defined. For example, the following is disallowed:
+
+```fsharp
+let test2 () =
+    let x = 12
+    &x // Error: 'y' escapes its defined scope!
+
+let test () =
+    let x =
+        let y = 1
+        &y // Error: `y` escapes its defined scope!
+    ()
+```
+
+This is for soundness reasons, as it prevents compiling code that could crash at runtime.
```

Comment:
This wasn't true, it would produce different results if compiled with optimizations on and off. Besides soundness reasons, this just makes the optimization passes have guarantees that a let-bound value will never go out of its scope.

Reply:
Gotcha. I won't mention optimization passes, though. That's just an internal detail that doesn't help people.